### DISCUSSION TITLE
Ability to bind `ImageTexture` as uniform in global `RenderingDevice` compute shader.

### DISCUSSION THREAD

**Discussion by adamlesinski** [godotengine/godot-proposals#7956] [Engine Core]
I'd like to run a compute shader on the global `RenderingDevice` to generate a texture that is then used by a fragment shader, without copying the texture between CPU and GPU memory.

Here is the code that would do this in Godot 4.1:

```
var rd := RenderingServer.get_rendering_device()
var shader_file := load("res://Compute.glsl")
var shader_spirv: RDShaderSPIRV = shader_file.get_spirv()
var shader := rd.shader_create_from_spirv(shader_spirv)
	
var texture := ImageTexture.create_from_image(Image.create(512, 512, false, Image.FORMAT_RF))
var texture_rid := RenderingServer.texture_get_rd_texture(texture)
	
var uniform := RDUniform.new()
uniform.uniform_type = RenderingDevice.UNIFORM_TYPE_IMAGE
uniform.binding = 0
uniform.add_id(texture_rid)
var uniform_set := rd.uniform_set_create([uniform], shader, 0)
	
var pipeline := rd.compute_pipeline_create(shader)
var compute_list := rd.compute_list_begin()
rd.compute_list_bind_compute_pipeline(compute_list, pipeline)
rd.compute_list_bind_uniform_set(compute_list, uniform_set, 0)
rd.compute_list_dispatch(compute_list, 512 / 8, 512 / 8, 1)
rd.compute_list_end()

# Can't call rd.submit() or rd.sync() because the renderer is in charge of this.
# Would be ideal if this pipeline would run before the renderer does.

$MeshInstance3D.get_surface_override_material(0).set_shader_parameter("texture", texture)
```

I had to make one change to the engine to get this to work: Add the `RD::TEXTURE_USAGE_STORAGE_BIT` flag to the texture that is created in `TextureStorage::texture_2d_initialize`.

Could this flag always be present? Is there a downside to including this flag at all times?

For more flexibility, there could be an API that would allow users to create an ImageTexture from the RID of a texture created from `RenderingDevice`.

Or better yet, just have `ShaderMaterial::set_shader_parameter` accept some `RDTexture` that represents the RID returned from `RenderingDevice::texture_create`.

**Comment by Calinou** [comment]
This appears to be already tracked in https://github.com/godotengine/godot-proposals/issues/6964, if I'm understanding right.

**Reply by adamlesinski** [reply]
Yep, looks like #6964 covers my case!

**Reply by Calinou** [reply]
That proposal was implemented in https://github.com/godotengine/godot/pull/79288, which isn't in 4.2.dev5 but it's in the latest `master` branch.
